sound/soc/sof/intel/mtl.c

```diff
@@ -0,0 +1,920 @@
+	romdbgsts = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HfFLGPxQWy);
+	romdbgerr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HfFLGPxQWy_ERROR);
+
+	dev_err(sdev->dev, "error: ROM status: 0x%x, ROM error: 0x%x\n", fwsts, fwlec);
+	dev_err(sdev->dev, "error: ROM debug status: 0x%x, ROM debug error: 0x%x\n", romdbgsts,
+		romdbgerr);
+	romdbgsts = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HfFLGPxQWy + 0x8 * 3);
+	dev_err(sdev->dev, "ROM feature bit%s enabled\n", romdbgsts & BIT(24) ? "" : " not");
+}
+
+static int mtl_dsp_core_power_up(struct snd_sof_dev *sdev, int core_mask)
+{
+	unsigned int cpa;
+	u32 dspcxctl;
+	int ret;
+
+	/* Only the primary core can be powered up by the host */
+	if (core_mask != BIT(0))
+		return 0;
+
+	/* Program the owner of the IP & shim registers (10: Host CPU) */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE,
+				MTL_DSP2CxCTL_PRIMARY_CORE_OSEL,
+				0x2 << MTL_DSP2CxCTL_PRIMARY_CORE_OSEL_SHIFT);
+
+	/* enable SPA bit */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE,
+				MTL_DSP2CxCTL_PRIMARY_CORE_SPA_MASK,
+				MTL_DSP2CxCTL_PRIMARY_CORE_SPA_MASK);
+
+	/* Wait for unstable CPA read (1 then 0 then 1) just after setting SPA bit */
+	usleep_range(1000, 1010);
+
+	/* poll with timeout to check if operation successful */
+	cpa = MTL_DSP2CxCTL_PRIMARY_CORE_CPA_MASK;
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE, dspcxctl,
+					    (dspcxctl & cpa) == cpa, HDA_DSP_REG_POLL_INTERVAL_US,
+					    HDA_DSP_RESET_TIMEOUT_US);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: %s: timeout on MTL_DSP2CxCTL_PRIMARY_CORE read\n",
+			__func__);
+		return ret;
+	}
+
+	/* did core power up ? */
+	dspcxctl = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE);
+	if ((dspcxctl & MTL_DSP2CxCTL_PRIMARY_CORE_CPA_MASK)
+		!= MTL_DSP2CxCTL_PRIMARY_CORE_CPA_MASK) {
+		dev_err(sdev->dev, "error: power up core failed core_mask %xadspcs 0x%x\n",
+			core_mask, dspcxctl);
+		ret = -EIO;
+	}
+
+	return ret;
+}
+
+static int mtl_dsp_core_power_down(struct snd_sof_dev *sdev, int core_mask)
+{
+	u32 dspcxctl;
+	int ret;
+
+	/* Only the primary core can be powered down by the host */
+	if (core_mask != BIT(0))
+		return 0;
+
+	/* disable SPA bit */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE,
+				MTL_DSP2CxCTL_PRIMARY_CORE_SPA_MASK, 0);
+
+	/* Wait for unstable CPA read (1 then 0 then 1) just after setting SPA bit */
+	usleep_range(1000, 1010);
+
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR, MTL_DSP2CxCTL_PRIMARY_CORE, dspcxctl,
+					    !(dspcxctl & MTL_DSP2CxCTL_PRIMARY_CORE_CPA_MASK),
+					    HDA_DSP_REG_POLL_INTERVAL_US,
+					    HDA_DSP_PD_TIMEOUT * USEC_PER_MSEC);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: failed to power down primary core\n");
+
+	return ret;
+}
+
+static int mtl_dsp_cl_init(struct snd_sof_dev *sdev, int stream_tag, bool imr_boot)
```

Comment:
You are planning to add the actual IMR boot support at a later time?

Reply:
it's line 459 @ujfalusi

Reply:
> it's line 459 @ujfalusi

The code was added after @ujfalusi reviewed :)